delegates/src/main/java/com/revolut/recyclerkit/delegates/DiffAdapter.kt

```diff
@@ -25,22 +28,113 @@ import androidx.recyclerview.widget.DiffUtil
  *
  */
 
-class DiffAdapter(
-    delegatesManager: DelegatesManager = DelegatesManager()
+open class DiffAdapter(
+    delegatesManager: DelegatesManager = DelegatesManager(),
+    async: Boolean = true,
+    autoScrollToTop: Boolean = false
 ) : AbsRecyclerDelegatesAdapter(delegatesManager) {
 
-    private val differ: AsyncListDiffer<ListItem> = AsyncListDiffer(
-        AdapterListUpdateCallback(this),
-        AsyncDifferConfig.Builder(ListDiffCallback<ListItem>()).build()
-    )
+    private interface DifferDelegate {
+        val items: List<ListItem>
+        fun setItems(items: List<ListItem>, recyclerView: RecyclerView)
+    }
+
+    private val differDelegate = if (async) AsyncDifferDelegate(this, autoScrollToTop) else SyncDifferDelegate(this, autoScrollToTop)
+    protected var recyclerView = WeakReference<RecyclerView>(null)
+        private set
+
+    open val items: List<ListItem>
+        get() = differDelegate.items
+
+    override fun getItem(position: Int): ListItem = differDelegate.items[position]
+
+    open fun setItems(items: List<ListItem>) {
+        val rv = recyclerView.get() ?: error("Recycler View not attached")
+        differDelegate.setItems(items, rv)
+    }
+
+    override fun getItemCount() = differDelegate.items.size
+
+    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
+        super.onAttachedToRecyclerView(recyclerView)
+        this.recyclerView = WeakReference(recyclerView)
+    }
+
+    private class AsyncDifferDelegate(
+        adapter: RecyclerView.Adapter<*>,
+        private val autoScrollToTop: Boolean
+    ) : DifferDelegate {
+        private val differ: AsyncListDiffer<ListItem> = AsyncListDiffer(
+            AdapterListUpdateCallback(adapter),
+            AsyncDifferConfig.Builder(ListDiffItemCallback<ListItem>()).build()
+        )
+        override val items: List<ListItem>
+            get() = differ.currentList
+
+        override fun setItems(items: List<ListItem>, recyclerView: RecyclerView) {
+            val layoutManager = recyclerView.layoutManager
+
+            val firstVisiblePosition = if (autoScrollToTop && layoutManager is LinearLayoutManager) {
+                layoutManager.findFirstCompletelyVisibleItemPosition()
+            } else {
+                -1
+            }
+
+            differ.submitList(items) {
+                if (firstVisiblePosition == 0) {
+                    recyclerView.scrollToPosition(0)
+                }
+            }
+        }
+    }
+
+    private class SyncDifferDelegate(
+        private val adapter: RecyclerView.Adapter<*>,
+        private val autoScrollToTop: Boolean
+    ) : DifferDelegate {
+        override val items = ArrayList<ListItem>()
+
+        override fun setItems(items: List<ListItem>, recyclerView: RecyclerView) {
+            val (diffResult, newList) = calculateDiff(items)
```

Comment:
It's unnecessary to return pair of result + newList from `calculateDiff` method. `newList` and `items` properties reference the same list